### PR TITLE
move from juju/errors to pkg/errors

### PR DESCRIPTION
**Pull request by lysu**
## What have you changed? (mandatory)

ref #7125 #2622 

This PR contain 3 commit, and recommand review by commit

1) first commit add pingcap/errors vendor that modified with https://github.com/pingcap/errors/pull/2
2) second commit fix some terror testcase
3) last commit use `sed` script to replace import packge and `make fmt`[no need review]
```bash
find . -name "*.go" | xargs sed -i 's/github.com\/juju\/errors/github.com\/pingcap\/tidb\/errors/g'
find . -name "*.go" | xargs sed -i 's/errors.Trace(err).(\*errors.Err).Cause()/errors.Cause(errors.Trace(err))/g'
```
revert and replay last step if have any conflict

## What is the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

- unit tests
- integration tests, need work with https://github.com/pingcap/tidb-test/pull/617 https://github.com/pingcap/tidb-private-test/pull/10

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

no

## Does this PR affect tidb-ansible update? (mandatory)

no

## Does this PR need to be added to the release notes? (mandatory)

no

## Refer to a related PR or issue link (optional)

#7125 #2622

https://github.com/pingcap/tidb-test/pull/617

https://github.com/pingcap/tidb-private-test/pull/10

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/7151)

### CODE REVIEW

**Comment by gregwebs** [comment]
Is it possible to get this PR merged? I am looking at doing the same in pd. I am starting to add stack traces in this PR https://github.com/pingcap/pd/pull/1200

**Comment by gregwebs** [comment]
@zhexuany can you click to load the diff and comment on specifically what you find strange?
It seems like juju/errors should be removed from it now.

File: cmd/benchfilesort/main.go [hunk 1]
```diff
@@ -80,15 +80,15 @@ func encodeRow(b []byte, row *comparableRow) ([]byte, error) {
 	sc := &stmtctx.StatementContext{TimeZone: time.Local}
 	body, err = codec.EncodeKey(sc, body, row.key...)
 	if err != nil {
-		return b, errors.Trace(err)
+		return b, errors.WithStack(err)
```
Comment (disksing) [inline cmd/benchfilesort/main.go]: Why not simply `return b, err`?

Comment (lysu) [reply cmd/benchfilesort/main.go]: this is first step of replacement,  finally this case will be `b, err`

(1)replace all errors.Trace to errors.Withstack (2) delete the "middle call" of `errors.WithStack` 

we can not repalce `b, errors.Trace` directly, because same "root call" maybe introduce some standard errors or third-part lib error which is not `pkg error` and need to  do a wrap..

It's hard to use script to detect it's "middle call" or "root call", so replace them to WithStack, then manual delete middle one orz..

or @disksing  is any other good advise~?

Comment (disksing) [reply cmd/benchfilesort/main.go]: Reasonable and fair. Good luck.

Comment (gregwebs) [reply cmd/benchfilesort/main.go]: This seems like a deficiency in the design of pkg/errors that we could remedy by just adding a new function that will take a stack trace if it exists? This has been [mentioned by others](https://github.com/pkg/errors/issues/75).

See this error implementation where they coalesce stack traces when printing: https://github.com/upspin/upspin/blob/master/errors/debug.go#L26.
Also see this error package which merges stack traces every time they are being wrapped: 
https://github.com/srvc/fail/blob/master/annotators.go#L50

In my PR for improved error codes I am now searching for the deepest stack trace:
https://github.com/pingcap/pd/pull/1200/files#diff-f10e5c1ee6d26ee1746a0a9188a67d4dR32
That assumes use of pkg/errors, but there is also now srvc/fail and go-stack/stack, all with different definitions of what a stack trace are.

Comment (lysu) [reply cmd/benchfilesort/main.go]: thanks~ @gregwebs , it makes sense, I try to implement it again, but use bottom-up way mark error already has stack info~ 3712fc1eb4b24c5a439788c3b5a4ee22375f5193

**Comment by lysu** [comment]
@coocood @zz-jason  PTAL, it seem already to ship.

File: errors/stack.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
+package errors
```
Comment (jackysp) [inline errors/stack.go]: Need a licence here.

Comment (lysu) [reply errors/stack.go]: Done~add license file.

File: errors/errors.go [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+}
+
+// Annotate returns an error annotating err with a stack trace
+// at the point Annotate is called, and the supplied message.
+// If err is nil, Annotate returns nil.
+func Annotate(err error, message string) error {
+	if err == nil {
+		return nil
+	}
+	errWithStack, hasStack := err.(withStackAware)
```
Comment (coocood) [inline errors/errors.go]: We can extract a function like
```
func hasStack(err error) bool
```

**Comment by gregwebs** [comment]
For traceability (this would prevent accidentally missing the license our copyright), and testability, I think we need to fork the package on github and add it with `dep`.

The current stack de-duping doesn't search the entire chain. I created a fork of pkg/errors with an `AddStack` function that does do that search:
https://github.com/pkg/errors/compare/master...gregwebs:dedup-stack-traces?expand=1

**Comment by coocood** [comment]
@gregwebs 
Forked here https://github.com/pingcap/errors

File: Gopkg.toml [hunk 4]
```diff
@@ -92,3 +92,7 @@ required = ["github.com/golang/protobuf/jsonpb"]
 [[constraint]]
   branch = "master"
   name = "github.com/etcd-io/gofail"
+
+[[constraint]]
```
Comment (jackysp) [inline Gopkg.toml]: It's better to move the vendor change to another pr.

Comment (lysu) [reply Gopkg.toml]: recommand review by commits, 1st comment add vendor, 2nd replace import and remove vendor, 3rd fix some testcase.

**Comment by lysu** [comment]
@coocood @zz-jason @jackysp  Now, we use pingcap/errors and ready to review PTAL if free thx~

**Comment by lysu** [comment]
run-all-tests tidb-test=pr/617 tidb-private-test=pr/10

**Comment by gregwebs** [comment]
We are maintaining backwards compatibility with pkg/errors. So it could be better to import this as pkg/errors rather than pingcap/errors. Any other libs that use pkg/errors will end up using our version of errors.

**Comment by lysu** [comment]
@gregwebs do you mean we use dep's `source` configuration, just like this?

```
[[constraint]]
  name = "github.com/pkg/errors"
  revision = "f29524922bfee4bf03f9c426c716e639cac63ea9"
  source = "https://github.com/pingcap/errors.git"
```
I try it, but meet this

```
dep ensure
Solving failure: No versions of github.com/pingcap/pd met constraints:
	master: Could not introduce github.com/pingcap/pd@master, as it has a dependency on github.com/pkg/errors with constraint ^0.8.0, which has no overlap with existing constraint f29524922bfee4bf03f9c426c716e639cac63ea9 from (root)
	v2.0.5: Could not introduce github.com/pingcap/pd@v2.0.5, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v2.0.4: Could not introduce github.com/pingcap/pd@v2.0.4, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v2.0.3: Could not introduce github.com/pingcap/pd@v2.0.3, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v2.0.2: Could not introduce github.com/pingcap/pd@v2.0.2, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v2.0.1: Could not introduce github.com/pingcap/pd@v2.0.1, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v2.0.0: Could not introduce github.com/pingcap/pd@v2.0.0, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v1.0.8: Could not introduce github.com/pingcap/pd@v1.0.8, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v1.0.7: Could not introduce github.com/pingcap/pd@v1.0.7, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v1.0.6: Could not introduce github.com/pingcap/pd@v1.0.6, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v1.0.5: Could not introduce github.com/pingcap/pd@v1.0.5, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v1.0.3: Could not introduce github.com/pingcap/pd@v1.0.3, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v1.0.2: Could not introduce github.com/pingcap/pd@v1.0.2, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v1.0.0: Could not introduce github.com/pingcap/pd@v1.0.0, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v2.1.0-rc.1: Could not introduce github.com/pingcap/pd@v2.1.0-rc.1, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v2.1.0-beta: Could not introduce github.com/pingcap/pd@v2.1.0-beta, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v2.1.0-alpha: Could not introduce github.com/pingcap/pd@v2.1.0-alpha, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v2.0.0-rc.6: Could not introduce github.com/pingcap/pd@v2.0.0-rc.6, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v2.0.0-rc.5: Could not introduce github.com/pingcap/pd@v2.0.0-rc.5, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v2.0.0-rc.4: Could not introduce github.com/pingcap/pd@v2.0.0-rc.4, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v2.0.0-rc.3: Could not introduce github.com/pingcap/pd@v2.0.0-rc.3, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v2.0.0-rc.1: Could not introduce github.com/pingcap/pd@v2.0.0-rc.1, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v1.1.0-beta: Could not introduce github.com/pingcap/pd@v1.1.0-beta, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v1.1.0-alpha.1: Could not introduce github.com/pingcap/pd@v1.1.0-alpha.1, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	v1.1.0-alpha: Could not introduce github.com/pingcap/pd@v1.1.0-alpha, as it is not allowed by constraint master from project github.com/pingcap/tidb.
	master: Could not introduce github.com/pingcap/pd@master, as it has a dependency on github.com/pkg/errors with constraint ^0.8.0, which has no overlap with existing constraint f29524922bfee4bf03f9c426c716e639cac63ea9 from (root)
```
so we need make pd use pingcap version first? :rofl:

**Comment by gregwebs** [comment]
Looks like pd needs to not have a pegged version of pkg/errors. I will send a PR there to use pingcap/errors and not peg the version. We probably also need to add a release tag to pingcap/errors, I will create an issue for that.

Please take a look at my small PRs to pingcap/errors: https://github.com/pingcap/errors/pulls

**Comment by lysu** [comment]
run-all-tests tidb-test=pr/617

**Comment by lysu** [comment]
run-all-tests tidb-test=pr/617 tidb-private-test=pr/10

File: Gopkg.toml [hunk 5]
```diff
@@ -96,3 +96,9 @@ required = ["github.com/golang/protobuf/jsonpb"]
 [[constraint]]
   branch = "master"
   name = "github.com/pingcap/tipb"
+
+[[constraint]]
+  name = "github.com/pkg/errors"
```
Comment (zz-jason) [inline Gopkg.toml]: why not using `pingcap/errors` directly?

Comment (lysu) [reply Gopkg.toml]: from @gregwebs 's view keep using pkg/errors make better compatiility https://github.com/pingcap/tidb/pull/7151#issuecomment-419429364

Comment (zz-jason) [reply Gopkg.toml]: got it.

Comment (lysu) [reply Gopkg.toml]: the good point is came when tidb need a third-part libray and it using pkg/errors too, use this way, all code use a same errors code.

File: terror/terror.go [hunk 6]
```diff
@@ -218,25 +217,23 @@ func (e *Error) getMsg() string {
...
 	err.args = args
-	_, err.file, err.line, _ = runtime.Caller(1)
-	return &err
+	return errors.AddStack(&err)
 }
 
 // FastGen generates a new *Error with the same class and code, and a new formatted message.
 // This will not call runtime.Caller to get file and line.
-func (e *Error) FastGen(format string, args ...interface{}) *Error {
+func (e *Error) FastGen(format string, args ...interface{}) error {
```
Comment (zz-jason) [inline terror/terror.go]: 1. return `*Error`?
2. why not warping a `errors.AddStack()` function call?

File: store/tikv/backoff.go [hunk 7]
```diff
@@ -136,7 +136,7 @@ func (t backoffType) String() string {
 	return ""
 }
 
-func (t backoffType) TError() *terror.Error {
+func (t backoffType) TError() error {
```
Comment (zz-jason) [inline store/tikv/backoff.go]: `*terror.Error` ?

File: store/tikv/2pc.go [hunk 8]
```diff
@@ -621,7 +621,8 @@ func (c *twoPhaseCommitter) execute(ctx context.Context) error {
 	if err != nil {
 		if undeterminedErr := c.getUndeterminedErr(); undeterminedErr != nil {
 			log.Warnf("con:%d 2PC commit result undetermined, err: %v, rpcErr: %v, tid: %v", c.connID, err, undeterminedErr, c.startTS)
-			err = errors.Wrap(err, terror.ErrResultUndetermined)
+			log.Error(err)
+			err = errors.Trace(terror.ErrResultUndetermined)
```
Comment (zz-jason) [inline store/tikv/2pc.go]: can we remove the `errors.Trace()`?

File: executor/write.go [hunk 9]
```diff
@@ -180,10 +181,11 @@ func updateRecord(ctx sessionctx.Context, h int64, oldData, newData []types.Datu
 
 // resetErrDataTooLong reset ErrDataTooLong error msg.
 // types.ErrDataTooLong is produced in types.ProduceStrWithSpecifiedTp, there is no column info in there,
-// so we reset the error msg here, and wrap old err with errors.Wrap.
+// so we reset the error msg here, and wrap old err with new error.
 func resetErrDataTooLong(colName string, rowIdx int, err error) error {
 	newErr := types.ErrDataTooLong.Gen("Data too long for column '%v' at row %v", colName, rowIdx)
-	return errors.Wrap(err, newErr)
+	log.Error(err)
```
Comment (zz-jason) [inline executor/write.go]: remove the log?

File: executor/write.go [hunk 10]
```diff
@@ -180,10 +181,11 @@ func updateRecord(ctx sessionctx.Context, h int64, oldData, newData []types.Datu
 
 // resetErrDataTooLong reset ErrDataTooLong error msg.
 // types.ErrDataTooLong is produced in types.ProduceStrWithSpecifiedTp, there is no column info in there,
-// so we reset the error msg here, and wrap old err with errors.Wrap.
+// so we reset the error msg here, and wrap old err with new error.
```
Comment (zz-jason) [inline executor/write.go]: update the comment?

File: store/tikv/backoff.go (hunk 7 above)
Comment (lysu) [reply store/tikv/backoff.go]: woh, `terror` is one of two logic change in 3rd commit

in previous `terror.Error` maintain stack frame by itself via `file`, `line` https://github.com/pingcap/tidb/pull/7151/commits/07a6f2968a11114fe7cb4865a783b238f5e54b93#diff-33cf41394e16bb72bf075a6278c80d52L225, so `GenXXX` can return `*terror.Error` directly.

but after this patch, `GenXXX` will return a wrap of `terror.Error`(pkg's WithStack),  so we need return a error interface, and in test, we use `errors.Cause` to check.

File: terror/terror.go (hunk 6 above)
Comment (lysu) [reply terror/terror.go]: `FastGen` seems be used in some situation doesn't need any stack from comment and old code logic...maybe I need more check for them usage..

File: store/tikv/2pc.go (hunk 8 above)
Comment (lysu) [reply store/tikv/2pc.go]: for this point, maybe better keep `Trace` beacause ErrResultUndetermined is reusable global variable without stack.

File: executor/write.go (hunk 9 above)
Comment (gregwebs) [reply executor/write.go]: I think this is a problem with `ErrDataTooLong` not being a struct with an `error` field. But we should create the chain, and we can do it this way (this could be moved upstream to pingcap/errors):

``` go
type ErrorChain struct {
  newErr: error
  oldErr: error
}
func (e ErrorChain) Error() string { return newErr.Error() }
func (e ErrorChain) Unwrap() error { return oldErr }
```

Comment (gregwebs) [reply executor/write.go]: https://github.com/pingcap/errors/pull/6

Comment (gregwebs) [reply executor/write.go]: You could try uber-go/multierr.

**Comment by gregwebs** [comment]
LGTM: All my comments of issues have been properly addressed.

**Comment by zhexuany** [comment]
We have to resolve conflicts in order to merge into master.

File: executor/write.go (hunk 10 above)
Comment (lysu) [reply executor/write.go]: yes, it's no longer use errors.Wrap so change the comment

File: terror/terror.go (hunk 6 above)
Comment (lysu) [reply terror/terror.go]: I check FastGen's usage , it just use in duplicate key error situation which are no need stack info. @zz-jason

File: executor/executor_test.go [hunk 11]
```diff
@@ -990,12 +990,12 @@ func (s *testSuite) TestUnion(c *C) {
 
 	_, err := tk.Exec("select 1 from (select a from t limit 1 union all select a from t limit 1) tmp")
 	c.Assert(err, NotNil)
-	terr := errors.Trace(err).(*errors.Err).Cause().(*terror.Error)
+	terr := errors.Cause(errors.Trace(err)).(*terror.Error)
```
Comment (zz-jason) [inline executor/executor_test.go]: can be simplified to "errors.Cause(err).(*terror.Error)"?

File: terror/terror.go [hunk 12]
```diff
@@ -218,25 +217,23 @@ func (e *Error) getMsg() string {
 }
 
 // Gen generates a new *Error with the same class and code, and a new formatted message.
-func (e *Error) Gen(format string, args ...interface{}) *Error {
+func (e *Error) Gen(format string, args ...interface{}) error {
```
Comment (zz-jason) [inline terror/terror.go]: how about change the function name to **"GenWithStack"** and change "GenByArgs" to **"GenWithStackByArgs"**?

File: executor/executor_test.go (hunk 11 above)
Comment (zz-jason) [reply executor/executor_test.go]: It's OK to not change this code, just casually ask.

Comment (lysu) [reply executor/executor_test.go]: yes and it's simple to change~

**Comment by lysu** [comment]
/run-common-test tidb-test=pr/617 tidb-private-test=pr/10
/run-integration-common-test tidb-test=pr/617 tidb-private-test=pr/10

**Comment by lysu** [comment]
/run-common-test tidb-test=pr/617 tidb-private-test=pr/10
/run-integration-common-test tidb-test=pr/617 tidb-private-test=pr/10

**Comment by gregwebs** [comment]
:tada: thank you!
